src/frequenz/sdk/timeseries/_serializable_ringbuffer.py

```diff
@@ -0,0 +1,91 @@
+from typing import List, TypeVar
+
+import numpy as np
+import numpy.typing as npt
+
+from ._ringbuffer import OrderedRingBuffer
+
+Container = TypeVar("Container", List[float], npt.NDArray[np.float64])
+
+# Version of the file dumping/loading format
+FILE_FORMAT_VERSION: int = 1
+
+
+class SerializableRingBuffer(OrderedRingBuffer[Container]):
+    """Sorted ringbuffer with serialization support."""
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        buffer: Container,
+        sampling_period: timedelta,
+        path: str,
+        time_index_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """Initialize the time aware ringbuffer.
+
+        Args:
+            buffer: Instance of a buffer container to use internally.
+            sampling_period: Timedelta of the desired resampling period.
+            path: Path to where the data should be saved to and loaded from.
+            time_index_alignment: Arbitrary point in time used to align
+                timestamped data with the index position in the buffer.
+                Used to make the data stored in the buffer align with the
+                beginning and end of the buffer borders.
+                For example, if the `time_index_alignment` is set to
+                "0001-01-01 12:00:00", and the `sampling_period` is set to
+                1 hour and the length of the buffer is 24, then the data
+                stored in the buffer could correspond to the time range from
+                "2022-01-01 12:00:00" to "2022-01-02 12:00:00" (date chosen
+                arbitrarily here).
+        """
+        super().__init__(buffer, sampling_period, time_index_alignment)
+        self._path = path
+        self._file_format_version = FILE_FORMAT_VERSION
+
+    def dump(self) -> None:
+        """Dump data to disk.
+
+        Raises:
+            I/O related exceptions when the file cannot be written.
+        """
+        with open(self._path, mode="wb+") as fileobj:
+            pickle.dump(self, fileobj)
+
+    @classmethod
+    def load(cls, path: str) -> SerializableRingBuffer[Container] | None:
+        """Load data from disk.
+
+        Args:
+            path: Path to the file where the data is stored.
+
+        Raises:
+            I/O related exceptions when file exists but isn't accessable for any
+            reason.
+
+        Returns:
+            `None` when the file doesn't exist, otherwise an instance of the
+            `SerializableRingBuffer` class, loaded from disk.
+        """
+        if not exists(path):
+            return None
+
+        with open(path, mode="rb") as fileobj:
+            instance: SerializableRingBuffer = pickle.load(fileobj)
+            instance._path = path  # pylint: disable=protected-access
+            # Set latest file format version for next time it dumps.
+            # pylint: disable=protected-access
+            instance._file_format_version = FILE_FORMAT_VERSION
```

Comment:
While performance is important, I think it's not really an issue because:

* we're loading 4 million entries in 0.01 second, we are very fast.
* Pickle would just fail pretty early if it's not a pickle format
* we can try loading the new format first instead of v1-pickle and optimially that format allows us to check the version using the first few bytes.
* It would be a temporary state until we update all files and then we can just remove the old version.